### ISSUE TITLE
about qwen2.5+siglip+sa2va not work

### ISSUE THREAD

**Issue by lxrrrrrr** [magic-research/Sa2VA#76]
Dear Author of Sa2VA,

I recently came across your work and found it very interesting. As a result, I spent several days integrating the SAM component of Sa2VA into the LLaVA-OneVision framework. However, after training with your provided data, I encountered an issue: the Dice loss consistently fails to decrease, and the trained SAM model does not follow the segmentation prompts.

I made sure to correctly add the [SEG] token and used the appropriate hidden states, but SAM seems unable to receive any prompt information—almost as if something is blocking the flow of information from the VLM to SAM. Could this be due to computational differences between the hidden states of Qwen2.5 and InternLM?

Below is my code for reference. I would greatly appreciate your insights. Thank you!

`
def forward(
        self,
        input_ids: torch.LongTensor = None,
        attention_mask: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.LongTensor] = None,
        past_key_values: Optional[List[torch.FloatTensor]] = None,
        inputs_embeds: Optional[torch.FloatTensor] = None,
        labels: Optional[torch.LongTensor] = None,
        use_cache: Optional[bool] = None,
        output_attentions: Optional[bool] = None,
        output_hidden_states: Optional[bool] = True,
        images: Optional[torch.FloatTensor] = None,
        image_sizes: Optional[List[List[int]]] = None,
        return_dict: Optional[bool] = None,
        modalities: Optional[List[str]] = ["image"],
        dpo_forward: Optional[bool] = False,
        cache_position=None,
        time_embedding=None,
        sam_image=None,
        mask=None,
        path=None,
    ) -> Union[Tuple, CausalLMOutputWithPast]:

        input_ids_sam=input_ids

        if inputs_embeds is None:
            (input_ids, position_ids, attention_mask, past_key_values, inputs_embeds, labels) = self.prepare_inputs_labels_for_multimodal(input_ids, position_ids, attention_mask, past_key_values, labels, images, modalities, image_sizes,time_embedding)
        

        if dpo_forward:
            outputs = self.model(
                input_ids=input_ids,
                attention_mask=attention_mask,
                position_ids=position_ids,
                past_key_values=past_key_values,
                inputs_embeds=inputs_embeds,
                use_cache=use_cache,
                output_attentions=output_attentions,
                output_hidden_states=output_hidden_states,
                return_dict=return_dict,
            )

            hidden_states = outputs[0]
            logits = self.lm_head(hidden_states)
            return logits, labels

        else:
            outputs = super().forward(
                input_ids=input_ids,
                attention_mask=attention_mask,
                position_ids=position_ids,
                past_key_values=past_key_values,
                inputs_embeds=inputs_embeds,
                labels=labels,
                use_cache=use_cache,
                output_attentions=output_attentions,
                output_hidden_states=output_hidden_states,
                return_dict=return_dict,
            )
            
        #return outputs
    
        if mask is None:
            # require zero seg datas
            seg_valid = False
            sam_image, frames_per_batch, gt_masks = self._get_pesudo_data(
                dtype=self.torch_dtype,
                device=input_ids.device,
            )
        else:
            seg_valid = True

        ori_size_list = []
        
        frames_per_batch=[i.shape[0] for i in sam_image]
        
        #print(frames_per_batch)
        for i_bs, _mask in enumerate(mask):
            #print(_mask)
            mask_shape = _mask[0].shape[-2:]
            #print(mask_shape)
            ori_size_list += [mask_shape] * frames_per_batch[i_bs]
        

        seg_token_mask = labels == self.seg_token_idx
        #seg_token_mask=torch.roll(seg_token_mask, shifts=-1, dims=0)

        hidden_states = outputs.hidden_states
        
        hidden_states = self.text_hidden_fcs(hidden_states[-1])
        #print(seg_token_mask1)

        _zero = hidden_states.mean() * 0.0
        if seg_valid:
            pred_embeddings = hidden_states[seg_token_mask] + _zero
        else:
            pred_embeddings = hidden_states[:, :5].flatten(0, 1) + _zero
        
        seg_token_counts = seg_token_mask.int().sum(-1)
        
        if not seg_valid:
            seg_token_counts += 5
            
        # print(pred_embeddings.shape,seg_token_counts)
        # exit(0)
        
        pred_embeddings_list_ = torch.split(pred_embeddings, seg_token_counts.tolist(), dim=0)
        pred_embeddings_list = []
        
        
        for item in pred_embeddings_list_:
            if len(item) != 0:
                pred_embeddings_list.append(item)
                
                
        pred_embeddings_list_video, success = self.genetate_video_pred_embeddings(
            pred_embeddings_list, frames_per_batch)
        
        if not success:
            raise NotImplementedError
            # return {'llm_loss': output.loss, 'loss_mask': output.loss * 0.0, 'loss_dice': output.loss * 0.0}

        gt_masks_video = self.process_video_gt_masks(mask, frames_per_batch)

        
        pred_embeddings_list_video, gt_masks_video = self.check_obj_number(
            pred_embeddings_list_video, gt_masks_video,pred_embeddings_list_video[0].shape[0]
        )
        

        
    
        g_pixel_values = torch.stack([self.sam_model.preprocess_image(pixel) for pixel in sam_image])[0]
        #print(g_pixel_values.shape) torch.Size([1, 3, 1024, 1024])

        num_objs = pred_embeddings_list_video[0].shape[0]
        num_frames = len(pred_embeddings_list_video)
        
        
        language_embeddings = torch.cat(pred_embeddings_list_video, dim=0)[:, None]
        #print(language_embeddings.shape,end=' *') torch.Size([2, 1, 256])

        sam_states = self.sam_model.get_sam2_embeddings(g_pixel_values, expand_size=num_objs)

        pred_masks = self.sam_model.inject_language_embd(sam_states, language_embeddings, nf_nobj=(num_frames, num_objs))
        
        #print(pred_masks.shape,end=' &') #torch.Size([1, 5, 256, 256])
        #exit(0) torch.Size([1, 2, 256, 256]) &

        gt_masks = [F.interpolate(gt_mask.unsqueeze(0), size=pred_masks[0].shape[-2:], mode='nearest').squeeze(0) for gt_mask in gt_masks_video]
    
    
        #print(len(gt_masks),gt_masks[0].shape,end='*\n') 1 torch.Size([2, 256, 256])*
        gt_masks = torch.cat(gt_masks, dim=0).to(input_ids_sam.device)
        #print(gt_masks.shape,end='(\n') torch.Size([2, 256, 256])(
        pred_masks = pred_masks.flatten(0, 1).to(input_ids_sam.device)
        #print(pred_masks.shape,end='&\n') torch.Size([2, 256, 256])&
        
        bs = len(pred_masks)
        loss_mask, loss_dice = 0, 0
        if len(pred_masks) != len(gt_masks):
            # drop this data
            print(f"Pred mask shape {pred_masks.shape} is not equal to gt_mask shape {gt_masks.shape} !!!")
            min_num = min(len(pred_masks), len(gt_masks))
            pred_masks = pred_masks[:min_num]
            gt_masks = gt_masks[:min_num]
            seg_valid = False

        if self.loss_sample_points:
            sampled_pred_mask, sampled_gt_mask = self.sample_points(pred_masks, gt_masks)
            sam_loss_dice = self.loss_dice(
                sampled_pred_mask,
                sampled_gt_mask, avg_factor=(len(gt_masks) + 1e-4))
            sam_loss_mask = self.loss_mask(
                sampled_pred_mask.reshape(-1),
                sampled_gt_mask.reshape(-1),
                avg_factor=(pred_masks.shape[0] * sampled_pred_mask.shape[1] + 1e-4))
        else:
            sam_loss_mask = self.loss_mask(pred_masks, gt_masks)
            sam_loss_dice = self.loss_dice(pred_masks, gt_masks)

        loss_mask += sam_loss_mask
        loss_dice += sam_loss_dice

        if not seg_valid:
            _scale = 0.0
        else:
            _scale = 1.0
        loss_mask = loss_mask * _scale
        loss_dice = loss_dice * _scale

        if '0' in str(loss_mask.device):
            print('LLM loss: ',outputs['loss'].item(),'loss_mask: ',loss_mask.item(),'loss_dice : ',loss_dice.item())
        
        #outputs['loss']=0
        outputs['loss']+=loss_mask
        outputs['loss']+=loss_dice
        
        outputs['loss_mask']=loss_mask
        outputs['loss_dice']=loss_dice

        return outputs`

**Comment by HarborYuan** [comment]
Hi @lxrrrrrr ,

Thanks for your interest in our work.

I would like to point out that replacing LLM with LLaVA-onevision may not be a small stuff. You need to modify the template in the dataset processing, the way the vision encoder works in the model, and some gradient enabling details. I don’t think [SEG] part needs to be modified. So I think the problem may lie in the previous parts.
